extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/runtime/QuarkusPolicyEnforcer.java

```diff
@@ -0,0 +1,112 @@
+package io.quarkus.keycloak.pep.runtime;
+
+import static org.keycloak.representations.adapters.config.PolicyEnforcerConfig.EnforcementMode.DISABLED;
+
+import java.util.Collection;
+
+import org.keycloak.adapters.KeycloakDeployment;
+import org.keycloak.adapters.authorization.PathCache;
+import org.keycloak.adapters.authorization.PolicyEnforcer;
+import org.keycloak.common.util.PathMatcher;
+import org.keycloak.representations.adapters.config.AdapterConfig;
+import org.keycloak.representations.adapters.config.PolicyEnforcerConfig;
+
+public final class QuarkusPolicyEnforcer extends PolicyEnforcer {
+
+    private final boolean hasHandledDisabledPaths;
+    private final OfflinePathMatcher offlinePathMatcher;
+
+    QuarkusPolicyEnforcer(KeycloakDeployment deployment, AdapterConfig adapterConfig) {
+        super(deployment, adapterConfig);
+
+        // we want to prevent unnecessary remote call when getting path config
+        // that is only possible when enforcement is disabled for the path and there is no pattern
+        // as pattern requires call to Keycloak
+        if (!getEnforcerConfig().getLazyLoadPaths()) {
+            boolean anyPathHasPattern = false;
+            boolean anyPathEndsWithWildcard = false;
+            for (PolicyEnforcerConfig.PathConfig pathConfig : getPaths().values()) {
+                if (pathConfig.getPath() != null) {
+
+                    if (pathConfig.hasPattern()) {
+                        anyPathHasPattern = true;
+                        break;
+                    }
+
+                    if (pathConfig.getEnforcementMode() == DISABLED) {
```

Comment:
@michalvavrik Should this check be done alongside the null path check above, as the first check ?

Reply:
no, because I don't want to use this "offline" path matching if there is at least one pattern (disabled or enforced). My understanding of https://github.com/tpdgt/keycloak/blob/master/adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/authorization/PolicyEnforcer.java#L314 is that pattern must be resolved remotely.

Reply:
luckily, I added test `KeycloakAuthorizationOfflineEnforcerPatternTest` that would fail if someone had similar idea as you did.

Reply:
consider following cycles:
- pathconfig has mode enforcement, no pattern (WE CAN IGNORE)
- pathconfig has mode disabled and wildcard (we need to know, set flag)
- pathconfig has mode enforcement, path contains pattern (we need to know as we can't reliable determine if path is disabled path when at least one path contains config)

You can't return PERMIT if you don't know whether correct pathconfig has been resolved. Do you understand path matching code differently? It's quite complex, I might be wrong, but to me, it is risky to ignore patterns for enforcement paths.

Let's consider

ENFORCED  path /api/public/{info_id}
DISABLED path /api/public/*

Reply:
@michalvavrik If we have `ENFORCED path /api/public/{info_id}`, that was my question, why do we need to check if it has a pattern if it has to be enforced and the remote call is necessary anyway, whether it has a pattern or it has no pattern ? It does not matter what other paths config set, overlapping paths with wildcards, etc, what matters is that this particular path config is requiring an enforcement.

Reply:
You don't know whether request path has enforced mode until you resolve path config via path matching mechanism. When you path match URI that ends up being disabled, you can't simply ignore paths with enforcement pattern for same tenant as without remote call, you won't know if they are not better match.

Reply:
@michalvavrik Sure, this is not really config but I agree it is not really that can be easily detected, so then what happens if you have an identical, as far as matching is concerned, when only a single segment follows `api/public`:

```
ENFORCED path /api/public/*
DISABLED path /api/public/*
```

?

Reply:
You probably need to adjust a tiny bit a check (if path has a pattern or non-DISABLED ends with wildcard) - if true then local PERMIT is not possible

Reply:
That said, may be `*` is also treated as a pattern, so just adding one more test will clarify it

Reply:
> That said, may be `*` is also treated as a pattern, so just adding one more test will clarify it

IMO it should be according to keycloak docs example, but it is not as we know how the check is implemented, it only checks for `{`. I'll add a test though.

Reply:
@sberyozkin 
added test that checks `/api/conflict/*` that is both disabled and enforcing. The result is that enforcement is disabled, which pleases me (next best thing to validation failure, but IMO we should not do that, it's Keycloak job, also, it's not related to the PR)

Whether `*` is treated as a pattern is tested by the `KeycloakAuthorizationOfflineEnforcerTest` as if it were treated as a pattern by the Keycloak adapted, path matching would always be false and tests would fail.